### PR TITLE
ENH: scipy.interpolate.NearestNDInterpolator: cKDTree configurable

### PR DESCRIPTION
**Pull request by jor-**
Due to the changes at the `scipy.spatial.cKDTree` with the lastest scipy version (0.16.0) it is now possible to influence the construction of the kd-tree with some parameters (`compact_nodes` and `balanced_tree`). Both allow to put emphasis on faster build time or faster query time and lower memory usage.
Since the `scipy.spatial.cKDTree` is the basis of the `scipy.interpolate.NearestNDInterpolator`, this possibility should also exist in `scipy.interpolate.NearestNDInterpolator`. For this, the constructor of `scipy.interpolate.NearestNDInterpolator` has been expanded by these parameters.

### CODE REVIEW

**Comment by rgommers** [comment]
Looks to me like this change should be motivated by a cost-benefit analysis. The cost is that the use of a KD tree goes from an implementation detail to something that a user should understand, and the two parameters added are quite technical. How large is the benefit here?

**Comment by jor-** [comment]
My motivation for this extension was a problem by which I had to interpolate values for 4 million points with 150 million data points. The interpolation took some minutes with scipy 0.15 and roughly the same time with `compact_nodes=False` and `balanced_tree=False` at my extension for scipy 0.16. The current (unchangeable) setting in scipy 0.16 is `compact_nodes=True` and `balanced_tree=True`. With this, the interpolation took several hours for my example. I do not know if the statement can be generalized. But it shows that theses parameters highly affect the execution time.
Maybe it makes sense to abstract both parameters as one or at least set both to False internally in the interpolator.

**Comment by rgommers** [comment]
Okay, minutes to hours is convincing:)

I guess this is then also a performance regression in 0.16.x? Can't remember if we made the change on purpose. @sturlamolden do you have a better memory?

**Comment by rgommers** [comment]
If we're going to add this to the API, it may be better than to make everything configurable. ``leafsize`` for example may also be useful. How about a single argument ``kdtree_opts=dict()`` which takes all parameters to ``cKDTree.__init__`` in dict format?

**Comment by rgommers** [comment]
Regarding the current PR:
- style comment: lines need to be <80 char (PEP8). 
- new keyword(s) should have ``.. versionadded:: 0.17.0``

**Comment by sturlamolden** [comment]
Those parameters are expencted to have high effect on the execution time, depending on the data. The current defaults are similar to scikit-learn (```sklearn.neighbors.KDTree```), which was usually faster than ```scipy.spatial.cKDTree``` up to version 0.15.x. It is not possible to make a general statement about this. I set the defaults I thought would be the faster options in most cases.

Note that it does not affect the return values, only the size of the cKDTree object and the query time.

I would also think it would sometimes help to turn on multi-threading in the k-NN query. But again, this depends on the data.

**Comment by sturlamolden** [comment]
@jakevdp might also have some insights.

**Comment by ev-br** [comment]
I wonder if http://stackoverflow.com/questions/31819778/scipy-spatial-ckdtree-running-slowly is related. @sturlamolden what I really wonder is whether there's a way of giving users some heuristics. But this is OT for this specific PR :-).

**Comment by sturlamolden** [comment]
I am not sure if there is an easy way to make everyone happy. But it might also be a bad idea to expose inner details like this to the API of the interpolator.

**Comment by sturlamolden** [comment]
I am happy to change the default back to the sme as 0.15.x, but then someone else will complain about a performance regression. I don't know...

**Comment by sturlamolden** [comment]
Another thing is that the median is found by quickselect. So if the data is already sorted, it will behave very poorly. In this case it would actually help to shuffle the data before the kd-tree is constructed.

**Comment by ev-br** [comment]
Regarding the use in `interpolate`: I am strongly in favor of not hiding relevant options from user. Hence a big +1 to Ralf's suggestion for adding a `kdtree_opts` parameter (or just `options`,  whatever the name). IMO it's worth marking this PR as a backport-candidate, so that if scipy 0.16.1 becomes real, the kd-tree options are exposed. Note that I am not suggesting to change anything in KDTree itself, IMO the regression worth backporting is not exposing the config options.

Regarding the KDTree -- I do not have any intelligent suggestion of what the defaults should be. I can only imagine what would be helpful for me if I were a user. Ideally, there's some heuristics. Eg, "if the variance of your data is large, try using option X". If that's not possible, then just a big fat notice "try playing with parameters X, Y and Z, there are several orders of magnitude in performance either way" in the docstring/tutorial would be helpful.

**Comment by rgommers** [comment]
> I am happy to change the default back to the same as 0.15.x, but then someone else will complain about a performance regression. I don't know...

True. Probably it should only be changed again if it's really clear that one setting is better than the other. Although 500x is a very large slowdown....

**Comment by pv** [comment]
Some of the other interpolator classes allow passing in a Delaunay
instance as `points`. Allowing KDTree input for the class-based
interpolator could be better than adding `kdtree_opts`, then it's up to
the user to construct it if it helps in the special case.

**Comment by sturlamolden** [comment]
@rgommers I have a feeling that the problem is the partial sort becoming O(n**2). It is probably easy to create such a situation if a kd-tree is abused to find neighbors in a grid with the data points ordered. (A kd-tree should not be used for that, but many will probably do it out of lazyness.) If it is too easy we should perhaps consider to use ```balanced_tree=False``` as the default. Though it generally leads to slower queries, it shouldn't exhibit slowdowns of this magnitude.

**Comment by sturlamolden** [comment]
I don't think we have performance on sorted grid data in the benchmarks. But we probably should. As for ```NearestNDInterpolator``` it might make sence to default to ```balanced_tree=False``` because here it is actually likely that the kd-tree might be constructed from presorted (grid) data.

**Comment by jor-** [comment]
I also think that the best variant would be a single argument `kdtree_opts=dict()` like @rgommers suggested. If it is consent, I would change that. The default options would then be the default options of the kdtree. Moreover, I would add a note at the docs that these parameters may heavily affect the execution time.

**Comment by sturlamolden** [comment]
It seems the performance regression reported on stackoverflow was related to the data points originating from a grid, in which case the partial sort used to split nodes has quadratic time complexity instead of linear. We should probably do something about this issue, because grid data might be a common (ab)use case for kd-trees (and for ```NearestNDInterpolator``` it certainly is).

**Comment by sturlamolden** [comment]
Using introselect instead of quickselect would guarrantee O(n) performance. Currently we use quickselect similar to scikit-learn, but we could used modified introselect code from NumPy instead.

**Comment by ev-br** [comment]
@jor- Would you be interested in implementing either Ralf's suggestion of `options` dict or Pauli's suggestion of passing a `cKDTree` instance as x? Then it can go into scipy 0.17

**Comment by jor-** [comment]
Yes, I would prefer Ralf's suggestion with the `options` dict.
At Pauli's suggestion of passing a `cKDTree` instance, I see the following difficulty. Before the `NearestNDInterpolator` instantiates the `cKDTree`, the passed `points` are processed (for example scaled). If a `cKDTree` instance is passed this have to be done somehow afterwards. I am not sure if this can be done easily.
If everyone is okay with the `options` dict solution, I am going to implement this.

File: scipy/interpolate/ndgriddata.py [hunk 1]
```diff
@@ -41,8 +41,11 @@ class NearestNDInterpolator(NDInterpolatorBase):
         Rescale points to unit cube before performing interpolation.
         This is useful if some of the input dimensions have
         incommensurable units and differ by many orders of magnitude.
-
         .. versionadded:: 0.14.0
+    cKDTree_options : dict, optional
```
Comment (ev-br) [inline scipy/interpolate/ndgriddata.py]: I typically shun away from bikeshedding names, but this one is really suboptimal, there's no way I can type right from the first try. `tree_options` or just `options`?

File: scipy/interpolate/ndgriddata.py [hunk 2]
```diff
@@ -50,11 +53,11 @@ class NearestNDInterpolator(NDInterpolatorBase):
 
     """
 
-    def __init__(self, x, y, rescale=False):
+    def __init__(self, x, y, rescale=False, cKDTree_options=dict()):
```
Comment (ev-br) [inline scipy/interpolate/ndgriddata.py]: Please avoid using mutable default arguments. Better use `options=None`, and then `if options is None: options = dict()`.

**Comment by ev-br** [comment]
I'm surprised that KDTree does not have a rescale option. Maybe it should (not this PR anyway).

At any rate, this needs at least a smoke test. A benchmark showing the performance regression would be nice to have, but it's optional and can be done in a separate PR.

Also it's cleaner to rebase rather than merge master. As it stands, a single `git rebase upstream/master` is going to remove the merge commit.

Overall, I think it's OK as a regression fix (modulo nitpicks), but I'd ask @pv's to weigh on in.

**Comment by pv** [comment]
I'd note that the issue of not being able to rescale is the same in `LinearNDInterpolator`.
What is done in that case is that `rescale=True` together with a passed-in triangulation results to a ValueError.

File: scipy/interpolate/ndgriddata.py (hunk 1 above)
Comment (jor-) [inline scipy/interpolate/ndgriddata.py]: I changed it to `tree_options`.

File: scipy/interpolate/ndgriddata.py (hunk 2 above)
Comment (jor-) [inline scipy/interpolate/ndgriddata.py]: I changed it to `options=None`.

**Comment by jor-** [comment]
Yes, the same problem is in `LinearNDInterpolator`. You can not pass a triangulation result and use `rescale`.

**Comment by sturlamolden** [comment]
@ev-br A kd-tree is not scale invariant because it is only valid for Minkowski distances. Rescaling meas we must rebuild the whole tree. A ball-tree might be scale invariant, depending on the distance function (it is not restricted to Minkowski distances).

**Comment by ev-br** [comment]
Here's a branch where I squashed the commits (there's no reason for this to have three commits) and added a smoke test, https://github.com/ev-br/scipy/commits/pr/5094
Can you pull it over to this PR? This will make it easier to merge.

**Comment by jor-** [comment]
@ev-br: Your branch is pulled over.

**Comment by ev-br** [comment]
OK, rebased and merged in eeb9da67ae299c518f14651463a0a5c39903da30, aa2756df465a873046a0d2c34609c2ed5b0ace25 and 1d49ff0b06ac2e5358cdb7cdc2e1849530607b76.
Thank you @jor- !

(for future reference: I rebased on master and you merged my branch and this pulled in lots of extra stuff; we typically prefer rebasing in feature branches).
